dotnet-desktop-guide/net/winforms/printing/how-to-print-text-document.md

```diff
@@ -0,0 +1,53 @@
+---
+title: "How to: Print a Multi-Page Text File"
+ms.date: "04/28/2022"
+dev_langs: 
+  - "csharp"
+  - "vb"
+helpviewer_keywords: 
+  - "printing [Windows Forms], printing multiple pages"
+  - "text [Windows Forms], printing Windows Forms"
+  - "Windows Forms, printing text"
+  - "printing [Windows Forms], text"
+ms.custom: devdivchpfy22
+---
+# How to: Print a Multi-Page Text File in Windows Forms .NET
+
+It's common for Windows-based applications to print text. The <xref:System.Drawing.Graphics> class provides methods for drawing objects (graphics or text) to a device, such as a screen or printer. The following section describes in detail the process to print text file. This method doesn't support print of word and pdf files.
+
+> [!NOTE]
+> The <xref:System.Windows.Forms.TextRenderer.DrawText%2A> methods of <xref:System.Windows.Forms.TextRenderer> are not supported for printing. You should always use the <xref:System.Drawing.Graphics.DrawString%2A> methods of <xref:System.Drawing.Graphics>, as shown in the following code example, to draw text for printing purposes.
+
+## To print text
+
+01. Add a <xref:System.Drawing.Printing.PrintDocument> component and a string to your form.
+
+    :::code language="csharp" source="snippets/how-to-print-text-document/csharp/Form1.cs" id="add_string_to_your_form":::
+
+    :::code language="vb" source="snippets/how-to-print-text-document/vb/Form1.vb" id="add_string_to_your_form":::
+
+01. To print a document, set the <xref:System.Drawing.Printing.PrintDocument.DocumentName%2A> property to the document you wish to print. Then open and read the document content to the string you added previously.
+
+    :::code language="csharp" source="snippets/how-to-print-text-document/csharp/Form1.cs" id="set_DocumentName_and_string":::
+
+    :::code language="vb" source="snippets/how-to-print-text-document/vb/Form1.vb" id="set_DocumentName_and_string":::
+
+01. In the <xref:System.Drawing.Printing.PrintDocument.PrintPage> event handler, use the <xref:System.Drawing.Printing.PrintPageEventArgs.Graphics%2A> property of the <xref:System.Drawing.Printing.PrintPageEventArgs> class and the document contents to calculate line length and lines per page. After each page is drawn, check if it's the last page, and set the <xref:System.Drawing.Printing.PrintPageEventArgs.HasMorePages%2A> property of the <xref:System.Drawing.Printing.PrintPageEventArgs> accordingly. The <xref:System.Drawing.Printing.PrintDocument.PrintPage> event is raised until <xref:System.Drawing.Printing.PrintPageEventArgs.HasMorePages%2A> is `false`. Also, make sure the <xref:System.Drawing.Printing.PrintDocument.PrintPage> event is associated with its event-handling method.
```

Comment:
Scan the rest of this article and remove the duplicate XREFs. When the old article has many xrefs, it becomes distracting for people to read because it's a constant visual distraction of "i'm a link you could click me!" in the paragraphs. For example, PrintPage is listed three times, HasMorePages is twice, PrintPageEventArgs is twice, and Graphics is once.

![image](https://user-images.githubusercontent.com/67293991/165794925-dcd72739-12a4-4b70-9f7a-63bbbc914fe5.png)

Instead, keep the first references to the type and convert the others to code blocks

```suggestion
01. In the <xref:System.Drawing.Printing.PrintDocument.PrintPage> event handler, use the <xref:System.Drawing.Printing.PrintPageEventArgs.Graphics%2A> property of the <xref:System.Drawing.Printing.PrintPageEventArgs> class and the document contents to calculate line length and lines per page. After each page is drawn, check if it's the last page, and set the <xref:System.Drawing.Printing.PrintPageEventArgs.HasMorePages%2A> property of the `PrintPageEventArgs` accordingly. The `PrintPage` event is raised until `HasMorePages` is `false`. Also, make sure the `PrintPage` event is associated with its event-handling method.
```

additionally, not in my suggestion block above, the end of the paragraph mentions associating the `PrintPage` event, but that really should be step 2 above, telling them how to do it in Visual Studio:


01. Select the `PrintDocument` component in the Visual Designer. On the **Properties** pane, select the **Event** filter and and then double-click the `PrintPage` event to generate an event handler.


Something like that.